api/emoji.go

```diff
@@ -0,0 +1,239 @@
+// See License.txt for license information.
+
+package api
+
+import (
+	"bytes"
+	"image"
+	_ "image/gif"
+	_ "image/jpeg"
+	_ "image/png"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"strings"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/gorilla/mux"
+	"github.com/mattermost/platform/model"
+	"github.com/mattermost/platform/utils"
+)
+
+const (
+	MaxEmojiFileSize = 64 * 1024 // 64 KB
+	MaxEmojiWidth    = 128
+	MaxEmojiHeight   = 128
+)
+
+func InitEmoji() {
+	l4g.Debug(utils.T("api.emoji.init.debug"))
+
+	BaseRoutes.Emoji.Handle("/list", ApiUserRequired(getEmoji)).Methods("GET")
+	BaseRoutes.Emoji.Handle("/create", ApiUserRequired(createEmoji)).Methods("POST")
+	BaseRoutes.Emoji.Handle("/delete", ApiUserRequired(deleteEmoji)).Methods("POST")
+	BaseRoutes.Emoji.Handle("/{id:[A-Za-z0-9_]+}", ApiUserRequired(getEmojiImage)).Methods("GET")
+}
+
+func getEmoji(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !*utils.Cfg.ServiceSettings.EnableCustomEmoji {
+		c.Err = model.NewLocAppError("getEmoji", "api.emoji.disabled.app_error", nil, "")
+		c.Err.StatusCode = http.StatusNotImplemented
+		return
+	}
+
+	if result := <-Srv.Store.Emoji().GetAll(); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		emoji := result.Data.([]*model.Emoji)
+		w.Write([]byte(model.EmojiListToJson(emoji)))
+	}
+}
+
+func createEmoji(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !*utils.Cfg.ServiceSettings.EnableCustomEmoji {
+		c.Err = model.NewLocAppError("createEmoji", "api.emoji.disabled.app_error", nil, "")
+		c.Err.StatusCode = http.StatusNotImplemented
+		return
+	}
+
+	if !(*utils.Cfg.ServiceSettings.RestrictCustomEmojiCreation == model.RESTRICT_EMOJI_CREATION_ALL || c.IsSystemAdmin()) {
+		c.Err = model.NewLocAppError("createEmoji", "api.emoji.create.permissions.app_error", nil, "user_id="+c.Session.UserId)
+		c.Err.StatusCode = http.StatusUnauthorized
+		return
+	}
+
+	if len(utils.Cfg.FileSettings.DriverName) == 0 {
+		c.Err = model.NewLocAppError("createEmoji", "api.emoji.storage.app_error", nil, "")
+		c.Err.StatusCode = http.StatusNotImplemented
+		return
+	}
+
+	if r.ContentLength > MaxEmojiFileSize {
+		c.Err = model.NewLocAppError("createEmoji", "api.emoji.create.too_large.app_error", nil, "")
+		c.Err.StatusCode = http.StatusRequestEntityTooLarge
+		return
+	}
+
+	if err := r.ParseMultipartForm(MaxEmojiFileSize); err != nil {
+		c.Err = model.NewLocAppError("createEmoji", "api.emoji.create.parse.app_error", nil, err.Error())
+		c.Err.StatusCode = http.StatusBadRequest
+		return
+	}
+
+	m := r.MultipartForm
+	props := m.Value
+
+	emoji := model.EmojiFromJson(strings.NewReader(props["emoji"][0]))
+	if emoji == nil {
+		c.SetInvalidParam("createEmoji", "emoji")
+		return
+	}
+
+	// wipe the emoji id so that existing emojis can't get overwritten
+	emoji.Id = ""
+
+	// make sure the emoji data is okay before we spend time uploading the image when we don't need to
+	emoji.PreSave()
```

Comment:
That's a good point about the Id and PreSave, but I think it might be better to stay consistent with how PreSave updates the Id field since that's how it works for other model classes.

And I'm going to change the text of that comment since it's not really saying what I meant it to say. The actual reason I'm setting up the model object with `PreSave` and calling `IsValid` first is to do our best to validate the emoji before saving anything to the file system/database. That way, we avoid having to clean up orphaned files and database entries when common errors like missing fields or invalid names.